cpp/src/lists/contains.cu

```diff
@@ -0,0 +1,288 @@
+  operator()(cudf::detail::lists_column_device_view const& d_lists,
+             cudf::scalar const& search_key,
+             cudf::mutable_column_device_view output_bools,
+             rmm::cuda_stream_view stream,
+             rmm::mr::device_memory_resource* mr) const
+  {
+    assert(search_key.is_valid() && "skey should have been checked for nulls by this point.");
+    return op_impl<T>(d_lists, detail::make_pair_iterator<T>(search_key), output_bools, stream, mr);
+  }
+
+  template <typename T>
+  std::enable_if_t<cudf::is_numeric<T>() || std::is_same<T, cudf::string_view>::value, void>
+  operator()(cudf::detail::lists_column_device_view const& d_lists,
+             cudf::column_device_view const& d_search_keys,
+             bool search_keys_have_nulls,
+             cudf::mutable_column_device_view output_bools,
+             rmm::cuda_stream_view stream,
+             rmm::mr::device_memory_resource* mr) const
+  {
+    return search_keys_have_nulls
+             ? op_impl<T>(d_lists, d_search_keys.pair_begin<T, true>(), output_bools, stream, mr)
+             : op_impl<T>(d_lists, d_search_keys.pair_begin<T, false>(), output_bools, stream, mr);
+  }
+};
+
+}  // namespace
+
+namespace detail {
+
+std::unique_ptr<column> contains(cudf::lists_column_view const& lists,
+                                 cudf::scalar const& search_key,
+                                 rmm::cuda_stream_view stream,
+                                 rmm::mr::device_memory_resource* mr)
+{
+  using namespace cudf;
+  using namespace cudf::detail;
+
+  CUDF_EXPECTS(!cudf::is_nested(lists.child().type()),
+               "Nested types not supported in lists::contains()");
+  CUDF_EXPECTS(lists.child().type().id() == search_key.type().id(),
+               "Type of search key does not match list column element type.");
+  CUDF_EXPECTS(search_key.type().id() != type_id::EMPTY, "Type cannot be empty.");
+
+  auto const device_view = column_device_view::create(lists.parent(), stream);
+  auto const d_lists     = lists_column_device_view(*device_view);
+
+  rmm::device_buffer null_mask;
+  size_type num_nulls;
+
+  std::tie(null_mask, num_nulls) = construct_null_mask(
+    d_lists, search_key, lists.has_nulls() || lists.child().has_nulls(), stream, mr);
+
+  auto ret_bools = make_fixed_width_column(
+    data_type{type_id::BOOL8}, lists.size(), std::move(null_mask), num_nulls, stream, mr);
+
+  if (search_key.is_valid()) {
+    auto ret_bools_mutable_device_view =
+      mutable_column_device_view::create(ret_bools->mutable_view(), stream);
+
+    cudf::type_dispatcher(search_key.type(),
+                          lookup_functor{},
+                          d_lists,
+                          search_key,
+                          *ret_bools_mutable_device_view,
+                          stream,
+                          mr);
+  }
+
+  return ret_bools;
+}
+
+std::unique_ptr<column> contains(cudf::lists_column_view const& lists,
```

Comment:
Still feel like the two detail-level `contains()` could be combined...

Reply:
This bothers me as well. I'll take another crack at this tomorrow.

Reply:
Hmm... This hinges partly on combining the two `construct_null_mask()` functions. :/

Reply:
I've collapsed this into one function now.